spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCdcWithMerge.java

```diff
@@ -0,0 +1,169 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.extensions;
+
+import java.util.List;
+import java.util.Map;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.actions.ActionsProvider;
+import org.apache.iceberg.actions.Cdc;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.spark.Spark3Util;
+import org.apache.iceberg.spark.SparkCatalog;
+import org.apache.iceberg.spark.actions.SparkActions;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.catalyst.analysis.NoSuchTableException;
+import org.apache.spark.sql.catalyst.parser.ParseException;
+import org.junit.After;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runners.Parameterized;
+
+import static org.apache.iceberg.TableProperties.WRITE_DISTRIBUTION_MODE_NONE;
+import static org.apache.iceberg.spark.actions.BaseCdcSparkAction.RECORD_TYPE;
+
+public class TestCdcWithMerge extends SparkRowLevelOperationsTestBase {
+  public TestCdcWithMerge(String catalogName, String implementation, Map<String, String> config, String fileFormat,
+                          boolean vectorized, String distributionMode) {
+    super(catalogName, implementation, config, fileFormat, vectorized, distributionMode);
+  }
+
+  /**
+   * Overwrite the parameters in the parent class. We need different combinations.
+   */
+  @Parameterized.Parameters(name = "catalogName = {0}, implementation = {1}, config = {2}," +
+      " format = {3}, vectorized = {4}, distributionMode = {5}")
+  public static Object[][] parameters() {
+    return new Object[][]{
+        {"testhive", SparkCatalog.class.getName(),
+            ImmutableMap.of(
+                "type", "hive",
+                "default-namespace", "default"
+            ),
+            "parquet",
+            true,
+            WRITE_DISTRIBUTION_MODE_NONE
+        },
+        {"testhive", SparkCatalog.class.getName(),
+            ImmutableMap.of(
+                "type", "hive",
+                "default-namespace", "default"
+            ),
+            "parquet",
+            false,
+            WRITE_DISTRIBUTION_MODE_NONE
+        }
+    };
+  }
+
+  @Override
+  protected Map<String, String> extraTableProperties() {
+    return ImmutableMap.of(TableProperties.MERGE_MODE, "copy-on-write");
+  }
+
+  @BeforeClass
+  public static void setupSparkConf() {
+    spark.conf().set("spark.sql.shuffle.partitions", "4");
+  }
+
+  @After
+  public void removeTables() {
+    sql("DROP TABLE IF EXISTS %s", tableName);
+    sql("DROP TABLE IF EXISTS source");
+  }
+
+  private ActionsProvider actions() {
+    return SparkActions.get();
+  }
+
+  @Test
+  public void testMergeWithOnlyUpdateClause() throws ParseException, NoSuchTableException {
```

Comment:
@stevenzwu, Spark use pos deletes in the merge command.

Reply:
Are you say8ing Spark merge into writes pos deletes? if so, it doesn't answer my question why do we need to emit pos deletes, as pos deletes are already applied as delete filters when emit insert records for the same snapshot.

Reply:
As I mentioned here, https://github.com/apache/iceberg/pull/4539#discussion_r855609262, we only apply the pos deletes for data files generated in the same snapshot. We still emit pos deletes when they delete rows from previous snapshots.